lua/autorun/server/wall_kick_parkour.lua

```diff
@@ -1,139 +1,141 @@
-                    local rightVel = ply:GetRight() * ply:WorldToLocal( ply:GetVelocity() + ply:GetPos()).y
-                    local upVel = Vector( 0, 0, kHeight )
-                    if ply:GetVelocity().z > kHeight then
-                        upVel = Vector( 0, 0, ( ply:GetVelocity().z - kHeight ) )
-                    end
-                    
-                    ply:ViewPunch( Angle( 0, 0, 3 * -dir ) )
-                    
-                    local totalVel = rightVel + upVel + Vector( normal.x * kickDistance, normal.y * kickDistance, 0 )
-                    
-                    ply:SetVelocity( totalVel )
-                    
-                    local footstep = footstepSounds[ math.random( 1, #footstepSounds ) ]
-                    ply:EmitSound( footstep, 100, 100, 0.25 )
-                    
-                    local snd = softSounds[ math.random( 1, #softSounds ) ]
-                    ply:EmitSound( snd )
-                    
+        end
+
+        -- Average all of the trace hits
+        local normal = accumulatedNormal / hitCount
+
+        -- If that trace hit something, we can start the process of kicking off
+        if hit then
+
+            -- Round the normal to coarser increments to make sure we count very similar surfaces and avoid floating point imprecision
+            local roundTo = 10
+            local x = ( math.Round( ( normal.x * 100 ) / roundTo, 0 ) * roundTo ) / 100
+            local y = ( math.Round( ( normal.y * 100 ) / roundTo, 0 ) * roundTo ) / 100
+            local z = ( math.Round( ( normal.z * 100 ) / roundTo, 0 ) * roundTo ) / 100
+            local roundedNormal = Vector( x, y, z )
+
+            local kHeight = kickHeight or 275
+
+            if ( not ply.LastNormal or ply.LastNormal ~= roundedNormal ) then
+                local rightVel = ply:GetRight() * ply:WorldToLocal( ply:GetVelocity() + ply:GetPos() ).y
+                local upVel = Vector( 0, 0, kHeight )
+                if ply:GetVelocity().z > kHeight then
+                    upVel = Vector( 0, 0, ply:GetVelocity().z - kHeight )
                 end
-                
-                ply.LastNormal = roundedNormal
-                
+
+                ply:ViewPunch( Angle( 0, 0, 3 * -dir ) )
+
+                local totalVel = rightVel + upVel + Vector( normal.x * kickDistance, normal.y * kickDistance, 0 )
+
+                ply:SetVelocity( totalVel )
+
+                local footstep = footstepSounds[math.random( 1, #footstepSounds )]
+                ply:EmitSound( footstep, 100, 100, 0.25 )
+
+                local snd = softSounds[math.random( 1, #softSounds )]
+                ply:EmitSound( snd )
+
             end
+
+            ply.LastNormal = roundedNormal
         end
-	end
-end)
+    end
+end )
 
 hook.Add( "OnPlayerHitGround", "WallKickHitGround", function( ply )
-	ply.WallKickJumpedAlready = nil
-	ply.LastNormal = nil
-end)
+    local plyTbl = ply:GetTable()
```

Comment:
`ply:GetTable().test` and `ply.test` are the same except the gettable skips a expensive `__index` call, the reason i only do it in `OnPlayerHitGround` is because that always gets called even when players didn't press space.
i'll probably do it everywhere though yeah